### PR TITLE
Escalate missing ArchUnit runtime dependency

### PR DESCRIPTION
**Pull request by mweirauch**
Encountered during `0.5.0-SNAPSHOT` "testing".

EDIT:
* Was causing CNFE for com.tngtech.archunit.core.importer.ImportOption.
* Perhaps having archunit an explicit dependency in spring-modulith-runtime could help here a bit? (Don't have much time to check ATM.)

### CODE REVIEW

**Comment by odrotbohm** [comment]
A bit of wild guessing here, but it seems that the only way this can happen is by you having defined ArchUnit defined in test scope (which I think most folks that previously used ArchUnit will have it declared) for your project, right? That would trump the ArchUnit compile scope defined dependency coming from Spring Modulith Core.

In other words, I agree we should declare this explicitly in the runtime module. I just wanted to understand how you could actually get into the problematic situation in the first place.

**Review by odrotbohm — CHANGES_REQUESTED** [review]
Please rather declare the ArchUnit dependency explicitly.

**Comment by odrotbohm** [comment]
Requested changes just for reference. Drop me a note if I should quickly tweak the change to go with the POM adjustments. I'm fine either way.

**Comment by mweirauch** [comment]
Yes, I had `archunit` as a test scope dependency present before adding `spring-modulith-starter-jdbc` to project.

I added `archunit` to `spring-modulith-runtime` as an explicit dependency as you proposed. But given the case others might run into the same scenario, the test-scoped `archunit` would still win due to the "distance to the dependency root". (As does for me.) So I would propose to keep the safeguard checking for `archunit` presence on the auto configuration class just in case if that is fine with you.

**Comment by odrotbohm** [comment]
You're correct. In this case, I wonder whether we should really only conditionally apply the configuration (to prevent it from raising the CNF error), or if we should rather actively check for the presence of the dependency and throw an explaining exception (maybe through a configuration class annotated with `@ConditionalOnMissingClass` with a simple constructor always throwing an exception?).

**Comment by mweirauch** [comment]
Is it just my gut feeling that throwing an exception in an AutoConfiguration feels a little too much? I am not experienced enough in this area, though. Would an informal INFO-level warning be perhaps a little bit more polite?

Taking on step back: (Doesn't solve the question though)

Shouldn't the `-runtime` be opt-in anyway? I was actually not assuming `spring-modulith-starter-core` pulling in `spring-modulith-runtime`. Wouldn't `spring-modulith-starter-insight` as the "runtime support starter" be a better place to attract it? (Of course, `-actuator` and `-observability` allready attract it.)

**Comment by odrotbohm** [comment]
It's the other way round: `core` has the general module model buildup logic. Primary usage is through the `test` module, that pulls `core` in (usually `test` scoped for the user app). `runtime` has some configuration that bootstraps a `core` model at runtime for example to invoke `ApplicationModuleInitializers` (defined in `api`) in the topologically right order. Both `actuator` and `observability` in turn refer to `runtime` to back both the actuator endpoints and find the beans to instrument for observation.

Long story short, if you have `runtime` on your classpath, you *will* need ArchUnit on the runtime classpath, too. That's why I think rejecting the arrangement makes sense. The auto-configuration backing off will just cause other configuration to be able to access an `ApplicationRuntime`. I spoke to the Boot team, and they seemed fine with the exception approach, but recommended to also register a `FailureAnalyzer` to describe the problem in a more user-friendly way.

**Comment by mweirauch** [comment]
Something like this? (Pending tests! Haven't written such since the Spring Boot Micrometer Jersey AutoConfiguration ...)

**Comment by odrotbohm** [comment]
Looks good to me. I can take it from here 🙃.

**Comment by odrotbohm** [comment]
That's polished and merged.
